From 6e63ee297ddd5195e547231abea48407d5b1de00 Mon Sep 17 00:00:00 2001
From: Dan O'Donovan <danielx.o'donovan@intel.com>
Date: Thu, 29 May 2014 16:27:22 +0200
Subject: [PATCH 333/429] gpio: increase system limit of 256 GPIOs for Edison
 platform

The kernel GPIO subsystem imposes a default limit of 256 GPIO pins
for the system.  However, the Edison Arduino platform has 262 GPIOs.
This patch follows the arch/arm example to specify a higher limit for
specific platforms.

Signed-off-by: Dan O'Donovan <danielx.o'donovan@intel.com>
---
 arch/x86/Kconfig                       | 11 +++++++++++
 arch/x86/configs/i386_edison_defconfig |  2 ++
 arch/x86/include/asm/gpio.h            | 23 +++++++++++++++++++++++
 3 files changed, 36 insertions(+)

diff --git a/arch/x86/Kconfig b/arch/x86/Kconfig
index c7ec1b865888..aee32eedc1af 100644
--- a/arch/x86/Kconfig
+++ b/arch/x86/Kconfig
@@ -450,6 +450,7 @@ config X86_INTEL_MID
 	select SFI
 	select INTEL_SCU_IPC
 	select X86_PLATFORM_DEVICES
+	select ARCH_HAVE_CUSTOM_GPIO_H
 	---help---
 	  Intel MID is Intel's Low Power Intel Architecture (LPIA) based Mobile
 	  Internet Device(MID) platform.
@@ -759,6 +760,16 @@ config APB_TIMER
          as it is off-chip. APB timers are always running regardless of CPU
          C states, they are used as per CPU clockevent device when possible.
 
+config ARCH_NR_GPIO
+	int
+	depends on ARCH_HAVE_CUSTOM_GPIO_H
+	default 512 if X86_INTEL_MID
+	default 0
+	help
+	  Maximum number of GPIOs in the system.
+
+	  If unsure, leave the default value.
+
 # Mark as expert because too many people got it wrong.
 # The code disables itself when not needed.
 config DMI
diff --git a/arch/x86/configs/i386_edison_defconfig b/arch/x86/configs/i386_edison_defconfig
index 0b99257b1ab8..1a69c25aea28 100644
--- a/arch/x86/configs/i386_edison_defconfig
+++ b/arch/x86/configs/i386_edison_defconfig
@@ -380,6 +380,7 @@ CONFIG_CPU_SUP_TRANSMETA_32=y
 CONFIG_CPU_SUP_UMC_32=y
 # CONFIG_HPET_TIMER is not set
 # CONFIG_APB_TIMER is not set
+CONFIG_ARCH_NR_GPIO=512
 CONFIG_DMI=y
 CONFIG_NR_CPUS=2
 CONFIG_SCHED_SMT=y
@@ -1624,6 +1625,7 @@ CONFIG_PTP_1588_CLOCK=y
 # Enable PHYLIB and NETWORK_PHY_TIMESTAMPING to see the additional clocks.
 #
 # CONFIG_PTP_1588_CLOCK_PCH is not set
+CONFIG_ARCH_HAVE_CUSTOM_GPIO_H=y
 CONFIG_ARCH_WANT_OPTIONAL_GPIOLIB=y
 CONFIG_GPIO_DEVRES=y
 CONFIG_GPIOLIB=y
diff --git a/arch/x86/include/asm/gpio.h b/arch/x86/include/asm/gpio.h
index b3799d88ffcf..131c96d46901 100644
--- a/arch/x86/include/asm/gpio.h
+++ b/arch/x86/include/asm/gpio.h
@@ -1,4 +1,27 @@
+#ifndef _ARCH_X86_GPIO_H
+#define _ARCH_X86_GPIO_H
+
+#if CONFIG_ARCH_HAVE_CUSTOM_GPIO_H
+
+#if CONFIG_ARCH_NR_GPIO > 0
+#define ARCH_NR_GPIOS CONFIG_ARCH_NR_GPIO
+#endif
+
+#include <asm-generic/gpio.h>
+
+/* The trivial gpiolib dispatchers */
+#define gpio_get_value	__gpio_get_value
+#define gpio_set_value	__gpio_set_value
+#define gpio_cansleep	__gpio_cansleep
+#define gpio_to_irq	__gpio_to_irq
+
+#else /* ! CONFIG_ARCH_HAVE_CUSTOM_GPIO_H */
+
 #ifndef __LINUX_GPIO_H
 #warning Include linux/gpio.h instead of asm/gpio.h
 #include <linux/gpio.h>
 #endif
+
+#endif /* ! CONFIG_ARCH_HAVE_CUSTOM_GPIO_H */
+
+#endif /* _ARCH_X86_GPIO_H */
-- 
2.37.3

